middle_end/flambda2/terms/validity.mli

```diff
@@ -0,0 +1,67 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           *)
+(*            Mark Shinwell and Xavier Clerc, Jane Street Europe          *)
+(*                                                                        *)
+(*   Copyright 1996 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   Copyright 2017--2019 Jane Street Group LLC                           *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** {1 Validity of expressions in Flambda}
+
+    Some operations/primitives (and therefore by extension expressions)
+    can be valid only after certain points in the control flow of a program.
+    For instance, the most obvisous is probably an array_load of an
+    **immutable** array: the load is only valid after check on the length,
+    or rather the behaviour of the generated code will change depending on
+    whether the load is before or after the array length check, even though
+    since this is for **immutable** arrays, neither the load nor the length
+    access have any effects or coeffects. This shows instances where the
+    regular notion of effects and co-effects is not enough to completely
+    define which ordering of operations are equivalent.
+
+    Therefore, we add a new notion to the effects and coeffects (and placement)
+    bundle that is used by Flambda2 and more importantly to_cmm to decide
+    which re-ordering of expressions are valid. *)
+
+(** Validity of operations/expressions *)
+type t =
+  | Always_valid  (** Operations that are always valid *)
+  | Valid_after_some_branch
```

Comment:
Maybe `Can't_move_before_any_branch`
and `Can_move_anywhere` for the `Always_valid`

Reply:
Seems good to me, do you have any suggestion for the name of the module ?